src/ensembl/src/shared/autosuggest-search-field/AutosuggestSearchField.tsx

```diff
@@ -0,0 +1,230 @@
+    // move to the next group in the list
+    return [groupIndex + 1, 0];
+  } else {
+    return null; // should never happen, but makes Typescript happy
+  }
+}
+
+function getPreviousItemIndex(
+  props: Props,
+  currentItemIndex: MatchIndex | null
+): MatchIndex {
+  const { matchGroups, allowRawInputSubmission } = props;
+  const [groupIndex, itemIndex] = currentItemIndex || [null, null];
+  const lastGroupIndex = matchGroups.length - 1;
+  const lastGroupItemIndex = matchGroups[lastGroupIndex].matches.length - 1;
+  const lastItemIndex: MatchIndex = [lastGroupIndex, lastGroupItemIndex];
+  if (itemIndex === null) {
+    return lastItemIndex;
+  } else if (itemIndex > 0) {
+    // move to the previous item
+    return [groupIndex, itemIndex - 1] as MatchIndex;
+  } else if (groupIndex === 0) {
+    // this is the first item in the first group;
+    // either return null if submitting raw input is allowed,
+    // or cycle back to the very last item in the list
+    return allowRawInputSubmission ? null : lastItemIndex;
+  } else if (typeof groupIndex === 'number') {
+    // move to the last item in the previous group
+    const previousGroupIndex = groupIndex - 1;
+    const lastItemIndex = matchGroups[previousGroupIndex].matches.length - 1;
+    return [previousGroupIndex, lastItemIndex];
+  } else {
+    return null; // should never happen, but makes Typescript happy
+  }
+}
+
+const AutosuggestSearchField = (props: Props) => {
+  const initialHighlightedItemIndex: MatchIndex = props.allowRawInputSubmission
+    ? null
+    : [0, 0];
+  const [highlightedItemIndex, setHighlightedItemIndex] = useState(
+    initialHighlightedItemIndex
+  );
+  const [isSelected, setIsSelected] = useState(false);
+  const [canShowSuggesions, setCanShowSuggestions] = useState(true);
+  const element = useRef<HTMLDivElement>(null);
+
+  useEffect(() => {
+    setIsSelected(false);
+  }, [props.search]);
+
+  useEffect(() => {
+    const onClickOutside = (event: MouseEvent) => {
+      const currentElement = element.current;
+      if (!currentElement) return;
+
+      if (
+        event.target !== currentElement &&
+        !currentElement.contains(event.target as HTMLElement)
+      ) {
+        setCanShowSuggestions(false);
+      }
+    };
+    window.addEventListener('click', onClickOutside);
+    return () => window.removeEventListener('click', onClickOutside);
+  }, []);
+
+  const handleSelect = (match: any) => {
+    setIsSelected(true);
+    props.onSelect(match);
+  };
+
+  const handleBlur = () => {
+    props.onBlur && props.onBlur();
+  };
+
+  const handleFocus = () => {
+    if (!canShowSuggesions) {
+      setCanShowSuggestions(true);
+    }
+    props.onFocus && props.onFocus();
+  };
+
+  const handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
+    if ([keyCodes.UP, keyCodes.DOWN].includes(event.keyCode)) {
+      event.preventDefault();
+    }
+    if (event.keyCode === keyCodes.UP) {
```

Comment:
Could this `if else` block be moved to be within the scope of the previous `if` block so that we can avoid checking for two additional conditions when the keys other than `up` and `down` are pressed?

Reply:
It could, but then we'll have nested if-statements, which some developers dislike. I am happy to move it there if no-one minds.

Alternatively, of course, the first check could be the reverse of what it currently is:

```
if (![keyCodes.UP, keyCodes.DOWN].includes(event.keyCode)) {
      return;
}
event.preventDefault();

if (event.keyCode === keyCodes.UP) {
...
} else if (event.keyCode === keyCodes.DOWN) {
   setHighlightedItemIndex(getNextItemIndex(props, highlightedItemIndex));
}
```

Whichever you prefer.

@imran-ebi any preferences?

Reply:
This approach would be even better. 
Also, we won't need `if (event.keyCode === keyCodes.DOWN)`

Reply:
Updated.

Reply:
I too don't like nested if-statements much, but I am still fine if they are used.